mono/metadata/icall.c

```diff
@@ -8564,12 +8567,26 @@ void
 mono_icall_cleanup (void)
 {
 	g_hash_table_destroy (icall_hash);
-	g_hash_table_destroy (icall_hash_foreign);
 	g_hash_table_destroy (jit_icall_hash_name);
 	g_hash_table_destroy (jit_icall_hash_addr);
 	mono_os_mutex_destroy (&icall_mutex);
 }
 
+static void
+add_internal_call_with_flags (const char *name, gconstpointer method, guint32 flags)
+{
+	mono_icall_lock ();
+
+	MonoIcallHashTableValue *value = g_new (MonoIcallHashTableValue, 1);
```

Comment:
`g_new` should be before `mono_icall_lock`, to reduce lock hold times.

Reply:
`g_strdup` should also be outside of lock.

Reply:
Yes, the g_strdup was in the lock in old solution as well, but moving both out should definetly be done to reduce lock times. Will update.